File: StickerMaker.java
// StickerMaker.java
// Sticker Maker на Java (использует простой алгоритм, без внешних библиотек)

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class StickerMaker {
    public static void main(String[] args) throws IOException {
        String input = null;
        String output = null;
        String colorStr = null;
        int threshold = 30;
        boolean verbose = false;

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "-o":
                case "--output":
                    output = args[++i];
                    break;
                case "-c":
                case "--color":
                    colorStr = args[++i];
                    break;
                case "-t":
                case "--threshold":
                    threshold = Integer.parseInt(args[++i]);
                    break;
                case "-v":
                case "--verbose":
                    verbose = true;
                    break;
                case "-h":
                case "--help":
                    System.out.println("Использование: java StickerMaker [опции] <входной_файл>\n" +
                            "  -o, --output <файл>   Выходной файл (PNG)\n" +
                            "  -c, --color <HEX>     Цвет фона\n" +
                            "  -t, --threshold <N>   Порог (0-100)\n" +
                            "  -v, --verbose         Подробный вывод\n" +
                            "  -h, --help            Справка");
                    System.exit(0);
                default:
                    if (input == null) input = args[i];
            }
        }

        if (input == null) {
            System.err.println("Ошибка: не указан входной файл.");
            System.exit(1);
        }

        File inputFile = new File(input);
        if (!inputFile.exists()) {
            System.err.println("Ошибка: файл " + input + " не найден");
            System.exit(1);
        }

        if (output == null) {
            output = input.replaceAll("\\.[^.]+$", "") + "_sticker.png";
        }

        BufferedImage src = ImageIO.read(inputFile);
        int w = src.getWidth();
        int h = src.getHeight();

        // Определяем цвет фона
        Color bgColor = detectBackgroundColor(src);
        if (colorStr != null) {
            try {
                bgColor = Color.decode(colorStr);
            } catch (NumberFormatException e) {
                if (verbose) System.err.println("Предупреждение: неверный формат цвета, используем автоопределение");
            }
        }
        if (verbose) {
            System.out.printf("Цвет фона: RGB(%d,%d,%d)\n", bgColor.getRed(), bgColor.getGreen(), bgColor.getBlue());
        }

        // Создаём изображение с альфа-каналом
        BufferedImage dst = new BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB);
        double maxDist = Math.sqrt(3 * 255 * 255);
        double thresh = threshold / 100.0 * maxDist;

        for (int y = 0; y < h; y++) {
            for (int x = 0; x < w; x++) {
                int rgb = src.getRGB(x, y);
                Color c = new Color(rgb);
                double dist = colorDistance(c, bgColor);
                int alpha = (dist <= thresh) ? 0 : 255;
                dst.setRGB(x, y, (alpha << 24) | (rgb & 0x00FFFFFF));
            }
        }

        ImageIO.write(dst, "png", new File(output));
        if (verbose) {
            System.out.println("Фон удалён. Результат сохранён в " + output);
        }
    }

    private static Color detectBackgroundColor(BufferedImage img) {
        int w = img.getWidth();
        int h = img.getHeight();
        List<Color> pixels = new ArrayList<>();
        // Края
        for (int x = 0; x < w; x++) {
            pixels.add(new Color(img.getRGB(x, 0)));
            pixels.add(new Color(img.getRGB(x, h - 1)));
        }
        for (int y = 0; y < h; y++) {
            pixels.add(new Color(img.getRGB(0, y)));
            pixels.add(new Color(img.getRGB(w - 1, y)));
        }
        // Средний цвет
        long sr = 0, sg = 0, sb = 0;
        for (Color c : pixels) {
            sr += c.getRed();
            sg += c.getGreen();
            sb += c.getBlue();
        }
        int n = pixels.size();
        if (n == 0) return Color.WHITE;
        return new Color((int)(sr / n), (int)(sg / n), (int)(sb / n));
    }

    private static double colorDistance(Color c1, Color c2) {
        double dr = c1.getRed() - c2.getRed();
        double dg = c1.getGreen() - c2.getGreen();
        double db = c1.getBlue() - c2.getBlue();
        return Math.sqrt(dr*dr + dg*dg + db*db);
    }
}

File: StickerMaker.cs
// StickerMaker.cs
// Sticker Maker на C# (использует SixLabors.ImageSharp)

using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;
using SixLabors.ImageSharp.Processing;
using System;
using System.Collections.Generic;
using System.IO;

class StickerMaker
{
    static void Main(string[] args)
    {
        string input = null, output = null, colorStr = null;
        int threshold = 30;
        bool verbose = false;

        for (int i = 0; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "-o":
                case "--output":
                    output = args[++i];
                    break;
                case "-c":
                case "--color":
                    colorStr = args[++i];
                    break;
                case "-t":
                case "--threshold":
                    threshold = int.Parse(args[++i]);
                    break;
                case "-v":
                case "--verbose":
                    verbose = true;
                    break;
                case "-h":
                case "--help":
                    Console.WriteLine(@"Использование: dotnet run -- [опции] <входной_файл>
  -o, --output <файл>   Выходной файл (PNG)
  -c, --color <HEX>     Цвет фона
  -t, --threshold <N>   Порог (0-100)
  -v, --verbose         Подробный вывод
  -h, --help            Справка");
                    return;
                default:
                    if (input == null) input = args[i];
                    break;
            }
        }

        if (input == null)
        {
            Console.Error.WriteLine("Ошибка: не указан входной файл.");
            Environment.Exit(1);
        }
        if (!File.Exists(input))
        {
            Console.Error.WriteLine($"Ошибка: файл {input} не найден");
            Environment.Exit(1);
        }

        output = output ?? input.Replace(Path.GetExtension(input), "") + "_sticker.png";

        using (var image = Image.Load<Rgba32>(input))
        {
            int w = image.Width, h = image.Height;

            // Определяем цвет фона
            Rgba32 bgColor = DetectBackgroundColor(image);
            if (colorStr != null)
            {
                try
                {
                    bgColor = ColorHexToRgba(colorStr);
                }
                catch
                {
                    if (verbose) Console.Error.WriteLine("Предупреждение: неверный формат цвета, используем автоопределение");
                }
            }
            if (verbose)
            {
                Console.WriteLine($"Цвет фона: RGB({bgColor.R},{bgColor.G},{bgColor.B})");
            }

            double maxDist = Math.Sqrt(3 * 255 * 255);
            double thresh = threshold / 100.0 * maxDist;

            // Создаём новое изображение с альфа-каналом
            using (var dst = new Image<Rgba32>(w, h))
            {
                for (int y = 0; y < h; y++)
                {
                    for (int x = 0; x < w; x++)
                    {
                        Rgba32 px = image[x, y];
                        double dist = ColorDistance(px, bgColor);
                        byte alpha = (dist <= thresh) ? (byte)0 : (byte)255;
                        dst[x, y] = new Rgba32(px.R, px.G, px.B, alpha);
                    }
                }
                dst.SaveAsPng(output);
            }
        }

        if (verbose)
        {
            Console.WriteLine($"Фон удалён. Результат сохранён в {output}");
        }
    }

    private static Rgba32 DetectBackgroundColor(Image<Rgba32> img)
    {
        int w = img.Width, h = img.Height;
        var pixels = new List<Rgba32>();
        for (int x = 0; x < w; x++)
        {
            pixels.Add(img[x, 0]);
            pixels.Add(img[x, h - 1]);
        }
        for (int y = 0; y < h; y++)
        {
            pixels.Add(img[0, y]);
            pixels.Add(img[w - 1, y]);
        }
        long sr = 0, sg = 0, sb = 0;
        foreach (var p in pixels)
        {
            sr += p.R;
            sg += p.G;
            sb += p.B;
        }
        int n = pixels.Count;
        if (n == 0) return new Rgba32(255, 255, 255);
        return new Rgba32((byte)(sr / n), (byte)(sg / n), (byte)(sb / n));
    }

    private static double ColorDistance(Rgba32 c1, Rgba32 c2)
    {
        double dr = c1.R - c2.R;
        double dg = c1.G - c2.G;
        double db = c1.B - c2.B;
        return Math.Sqrt(dr * dr + dg * dg + db * db);
    }

    private static Rgba32 ColorHexToRgba(string hex)
    {
        hex = hex.TrimStart('#');
        if (hex.Length == 6)
        {
            byte r = Convert.ToByte(hex.Substring(0, 2), 16);
            byte g = Convert.ToByte(hex.Substring(2, 2), 16);
            byte b = Convert.ToByte(hex.Substring(4, 2), 16);
            return new Rgba32(r, g, b);
        }
        throw new FormatException("Неверный формат HEX цвета");
    }
}
